spec/formatting.md

```diff
@@ -320,16 +319,20 @@ Implementation-defined _functions_ SHOULD use an implementation-defined _namespa
 
 #### Option Resolution
 
-The result of resolving _option_ values is an unordered mapping of string identifiers to values.
+The result of resolving _option_ values is an unordered mapping of string identifiers to values,
+determined as follows:
 
-For each _option_:
+1. Let `res` be a new empty mapping.
+1. For each _option_:
+   1. Let `id` be the string value of the _identifier_ of the _option_.
+   1. Let `rv` be the _resolved value_ of the _option_ value.
+   1. If `rv` is a _fallback value_:
+      1. Set the value bound to `id` in `res` to an unresolved value.
```

Comment:
Is "unresolved value" defined somewhere? You're trying to avoid saying `null` right?

Reply:
It's not defined, and yes, this ought to be represented by something like `undefined` or `None` or a null pointer or something similar.

Reply:
Would this step result in unbinding `id`'s value if it already existed? Put another way, is there value to adding `id` to the mapping if the value is blank? Some environments might not tolerate null values as well. We might permit binding the unresolved value or not binding at all.

Reply:
I agree on avoiding "unresolved value". I would rather say something like "ensure no value is bound to `id` in `res`", because there are situations where there's no equivalent of null pointers/`undefined`. For example, in C++, the values in the map might not be pointers at all, or if they are, then binding a name to null might not be any different from not adding it to the map.

Reply:
For example:

```
.input {$foo :function option1=a option2=b}
.local $b = {$foo :function option1=$undeclaredVar}
{{{$b}}}
```

In think in this case what our spec has been trying to do is: the failure to resolve `$undeclaredVar` would result in `$b` having options `{option2: "b"}`, with `option1` not present (it's not `a` and it's not present-but-null). Either way you get a Bad Option error (if you want it). If `option1` were required, the _function handler_ would barf, but so to would it barf on a null value.

@eemeli are you thinking that you want to represent the "bad option" state into the data model?

Reply:
The term was originally introduced from a suggestion by @gibson042 in https://github.com/unicode-org/message-format-wg/pull/524#discussion_r1408097009, and subsequently discussed on the [2023-12-04 call](https://github.com/unicode-org/message-format-wg/blob/main/meetings/2023/notes-2023-12-04.md#should-we-remove-unresolved-options-from-the-map-545) (based on my recollection, but the notes don't cover that) and then amended in #545.

I would prefer not re-opening the discussion about what we're doing in general, but I'd be very open for suggestions about some better language to use here instead of "unreseolved value".

Note that if we require a _fallback value_ to be able to show up as a resolved _option_ value in _function handler_ arguments, then that'll introduce a novel requirement on implementations, as the formatting behaviour otherwise does not give them any public visibility.

Reply:
The purpose of "unresolved value" as originally suggested at https://github.com/unicode-org/message-format-wg/pull/524#discussion_r1408097009 is "a representation of failure to resolve [that is distinct from every possible resolved value]". If you're looking for a reasonable replacement, I would expect something like "a value uniquely representing failure to resolve" or "the _resolution failure value_" (with a corresponding definition).

Reply:
But that detail is very important IMO: it **must not** be possible for input to explicitly specify a value that the implementation would produce upon encountering an undeclared variable or similar, so that function implementations are guaranteed the ability to detect options suffering from such a state.

Reply:
> Finally, I'll note that "explicitly specified value that cannot be passed by the caller" is tricky, particularly in low-level language implementations. It suggests that the option values are boxed in some way so that the handler can check if it is an unresolved value (unresolved being disjoint from all data representations)

Absolutely, and my preference remains what it was in my original comment for #545—don't even let implementations see options whose value fails to resolve (_I accepted @stasm's argument about error messages to move things along, but I still think it would be better to rely only on the preceding Unresolved Variable error and present to functions only options that successfully resolve_).

But assuming options with unresolved values _are_ presented to functions, I don't consider the disjointness requirement to be negotiable.

Reply:
How about this, combined with dropping the bit below about removing unresolved values?
```suggestion
      1. Emit a _Bad Option_ error.
```